File: backend/app/schemas/purchase.py
from __future__ import annotations

from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import TimestampedSchema
from app.schemas.product import ProductRead
from app.schemas.supplier import SupplierRead
from app.schemas.user import UserRead


class PurchaseDetailCreate(BaseModel):
    product_id: UUID
    quantity: Decimal = Field(gt=0)
    unit_cost: Decimal = Field(ge=0)


class PurchaseCreate(BaseModel):
    supplier_id: UUID
    invoice_number: str | None = Field(default=None, max_length=60)
    tax_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    notes: str | None = None
    details: list[PurchaseDetailCreate] = Field(min_length=1)


class PurchaseDetailRead(TimestampedSchema):
    product_id: UUID
    quantity: Decimal
    unit_cost: Decimal
    line_total: Decimal
    product: ProductRead


class PurchaseRead(TimestampedSchema):
    supplier_id: UUID
    user_id: UUID
    invoice_number: str | None = None
    subtotal: Decimal
    tax_amount: Decimal
    total: Decimal
    notes: str | None = None
    supplier: SupplierRead
    user: UserRead
    details: list[PurchaseDetailRead]

File: frontend/src/features/purchases/types/purchase.ts
import type { Product } from "../../products/types/product";
import type { User } from "../../auth/types/auth";

export type Supplier = {
  id: string;
  name: string;
  tax_id?: string | null;
  phone?: string | null;
  email?: string | null;
  address?: string | null;
  is_active?: boolean;
};

export type PurchaseDetailPayload = {
  product_id: string;
  quantity: number;
  unit_cost: number;
};

export type PurchasePayload = {
  supplier_id: string;
  invoice_number?: string | null;
  tax_amount: number;
  notes?: string | null;
  details: PurchaseDetailPayload[];
};

export type PurchaseDetail = {
  id: string;
  product_id: string;
  quantity: string;
  unit_cost: string;
  line_total: string;
  created_at: string;
  updated_at: string;
  product: Product;
};

export type Purchase = {
  id: string;
  supplier_id: string;
  user_id: string;
  invoice_number?: string | null;
  subtotal: string;
  tax_amount: string;
  total: string;
  notes?: string | null;
  created_at: string;
  updated_at: string;
  supplier: Supplier;
  user: User;
  details: PurchaseDetail[];
};
